### PR TITLE
internal/cmd/gocdk: implement local launcher

### PR DESCRIPTION
**Pull request by zombiezen**
Adds full launch command.

Unsure what the testing strategy here should be, #2013 tracks that.

Updates #1861

### CODE REVIEW

File: internal/cmd/gocdk/launch.go [hunk 1]
```diff
@@ -15,11 +15,170 @@
...
+	// Read the launch specifier from the biome's Terraform output.
+	tfOutput, err := tfReadOutput(ctx,
+		findBiomeDir(moduleRoot, biome),
+		pctx.overrideEnv("TF_IN_AUTOMATION=1"))
+	if err != nil {
+		return xerrors.Errorf("gocdk launch: %w", err)
+	}
+
+	// Launch the application.
+	launchURL, err := launcher.Launch(ctx, &launchInput{
```
Comment (vangent) [inline internal/cmd/gocdk/launch.go]: How are you choosing which arguments to pass to `Launch` vs which ones are passed to `newLauncher`? It might be simpler to put them all in once place or the other...?

File: internal/cmd/gocdk/launch.go [hunk 2]
```diff
@@ -15,11 +15,170 @@
...
+	c.Dir = local.dockerDir
+	out, err := c.CombinedOutput()
+	if err != nil {
+		if len(out) == 0 {
+			return nil, xerrors.Errorf("local launch: docker run: %w", err)
+		}
+		return nil, xerrors.Errorf("local launch: docker run:\n%s", out)
+	}
+
+	local.logger.Printf("Docker container %s started, waiting for healthy...", bytes.TrimSuffix(out, []byte("\n")))
```
Comment (vangent) [inline internal/cmd/gocdk/launch.go]: How do they stop it?

File: internal/cmd/gocdk/launch.go [hunk 3]
```diff
@@ -15,11 +15,170 @@
...
+	serveURL := &url.URL{
+		Scheme: "http",
+		Host:   "localhost:8080",
+		Path:   "/",
+	}
+	healthCheckURL := &url.URL{
```
Comment (vangent) [inline internal/cmd/gocdk/launch.go]: Nit: Or, slightly shorter:

```
healthCheckURL:= *serveURL
healthCheckURL.Path = "/healthz/readiness"
```
?

File: internal/cmd/gocdk/launch.go [hunk 4]
```diff
@@ -15,11 +15,170 @@
...
+		Scheme: serveURL.Scheme,
+		Host:   serveURL.Host,
+		Path:   "/healthz/readiness",
+	}
+	if err := waitForHealthy(ctx, healthCheckURL); err != nil {
```
Comment (vangent) [inline internal/cmd/gocdk/launch.go]: Nit: would it be easier to use if `waitForHealthy` took an url string instead of a `url.URL`?

File: internal/cmd/gocdk/launch.go [hunk 5]
```diff
@@ -15,11 +15,170 @@
...
+		return nil, xerrors.Errorf("local launch: %w", err)
+	}
+	return serveURL, nil
+}
+
+// launchSpecifier is the set of arguments passed from a biome's Terraform
+// module to the biome's launcher.
+type launchSpecifier map[string]interface{}
```
Comment (vangent) [inline internal/cmd/gocdk/launch.go]: I am not sure that this type is adding much value. Consider leaving it as a `map[string]interface{}`. You can have still have a `stringValue` helper that takes one as an argument.

File: internal/cmd/gocdk/launch.go [hunk 6]
```diff
@@ -15,11 +15,170 @@
...
+	dockerDir string
+}
+
+// Launch implements Launcher.Launch.
+func (local *localLauncher) Launch(ctx context.Context, input *launchInput) (*url.URL, error) {
+	dockerArgs := []string{
+		"run",
+		"--rm",
+		"--detach",
+		"--publish", "8080:8080",
```
[the remark this answers is not in the document]
Comment (clausti) [reply internal/cmd/gocdk/launch.go]: it might be nice for it to be settable, but I think it's ok for running `serve` and `deploy` in the same biome simultaneously to be mutually exclusive.

[the remark this answers is not in the document]
Comment (zombiezen) [reply internal/cmd/gocdk/launch.go]: Added a launch specifier option.

File: internal/cmd/gocdk/launch.go (hunk 1 above)
Comment (zombiezen) [reply internal/cmd/gocdk/launch.go]: That which is specific to a launcher type and could be reused between launches goes into the launcher constructor. \
That which is generic to any launcher goes into the `launchInput`.

File: internal/cmd/gocdk/launch.go [hunk 7]
```diff
@@ -15,11 +15,170 @@
...
+	return nil
+}
+
+// Launcher is the interface for any type that can launch a Docker image.
+type Launcher interface {
+	Launch(ctx context.Context, input *launchInput) (*url.URL, error)
+}
+
+// launchInput is the input to a launcher.
+type launchInput struct {
```
[the remark this answers is not in the document]
Comment (zombiezen) [reply internal/cmd/gocdk/launch.go]: Done.

File: internal/cmd/gocdk/launch.go [hunk 8]
```diff
@@ -15,11 +15,170 @@
...
+	})
+	if err != nil {
+		return xerrors.Errorf("gocdk launch: %w", err)
+	}
+	fmt.Fprintf(pctx.stdout, "Serving at %s\n", launchURL)
+	return nil
+}
+
+// Launcher is the interface for any type that can launch a Docker image.
+type Launcher interface {
```
[the remark this answers is not in the document]
Comment (zombiezen) [reply internal/cmd/gocdk/launch.go]: TODO added.

File: internal/cmd/gocdk/launch.go [hunk 9]
```diff
@@ -15,11 +15,170 @@
...
+}
+
+// newLauncher creates the launcher for the given name.
+func newLauncher(ctx context.Context, pctx *processContext, launcherName string) (Launcher, error) {
+	logger := log.New(pctx.stderr, "gocdk: ", log.Ldate|log.Ltime)
+	switch launcherName {
+	case "local":
+		return &localLauncher{
+			logger:    logger,
+			dockerEnv: pctx.env,
```
[the remark this answers is not in the document]
Comment (zombiezen) [reply internal/cmd/gocdk/launch.go]: > It's confusing to me that `localLauncher` has a `dockerEnv` from `pctx.env`, but `launchInput` also has an `env`. What goes where? Can we just have one set of environment variables?

Based on discussions around #1890, using environment variables to pass configuration to the running binary seems right, so these are two distinct sets of environment variables. We can't join them.

> IIUC, one is the set used to exec `docker`, while the other is the set passed to the docker image being run. I guess that's reasonable. Maybe `localLauncher` can just have a `processContext` field instead of copying its subfields?

I was trying to be as specific as possible with what the `localLauncher` depends on from the `processContext`.

File: internal/cmd/gocdk/launch.go [hunk 10]
```diff
@@ -15,11 +15,198 @@
...
+	"path/filepath"
+	"strconv"
 
 	"golang.org/x/xerrors"
 )
 
 func launch(ctx context.Context, pctx *processContext, args []string) error {
-	return xerrors.New("not implemented")
+	f := newFlagSet(pctx, "launch")
+	dockerImage := f.String("image", "", "Docker image to launch in the form `image[:tag]` (defaults to project latest)")
```
Comment (clausti) [inline internal/cmd/gocdk/launch.go]: The dockermentation uses `name:tag`, and I think we should too. (Not objecting to the optional brackets, just to calling it "image" instead of "name".)

File: internal/cmd/gocdk/launch.go [hunk 11]
```diff
@@ -15,11 +15,170 @@
...
+	"os/exec"
+	"path/filepath"
 
 	"golang.org/x/xerrors"
 )
 
 func launch(ctx context.Context, pctx *processContext, args []string) error {
-	return xerrors.New("not implemented")
+	f := newFlagSet(pctx, "launch")
+	imageName := f.String("image", "", "Docker image name to launch (defaults to project latest)")
```
[the remark this answers is not in the document]
Comment (clausti) [reply internal/cmd/gocdk/launch.go]: Definitely, I think the biome/launcher concepts will be useful with non-gocdk projects. I'm cool with having users specify both the name and the tag, just not just the name.

File: internal/cmd/gocdk/launch.go [hunk 12]
```diff
@@ -15,11 +15,170 @@
...
+	}
+	launcher, err := newLauncher(ctx, pctx, *cfg.Launcher)
+	if err != nil {
+		return xerrors.Errorf("gocdk launch: %w", err)
+	}
+
+	// Read the launch specifier from the biome's Terraform output.
+	tfOutput, err := tfReadOutput(ctx,
+		findBiomeDir(moduleRoot, biome),
+		pctx.overrideEnv("TF_IN_AUTOMATION=1"))
```
[the remark this answers is not in the document]
Comment (vangent) [reply internal/cmd/gocdk/launch.go]: Shouldn't we set it in `tfReadOutput` then?

File: internal/cmd/gocdk/launch.go (hunk 2 above)
Comment (vangent) [reply internal/cmd/gocdk/launch.go]: Is `Launch` going to support an `Unlaunch` or `Stop` command? That would seem reasonable to me.

Comment (vangent) [reply internal/cmd/gocdk/launch.go]: (maybe file an issue for that).

And consider just extending this `Printf` with a note on how to stop the container.

File: internal/cmd/gocdk/launch.go (hunk 5 above)
Comment (vangent) [reply internal/cmd/gocdk/launch.go]: Sorry, I don't follow your comment. The extra type made this harder to follow for me, not easier. We'll probably be reading config variables from the json config in various places, I think it will be easier to follow if we just have a set of utilities for doing so rather than adding a new type for each use case.

File: internal/cmd/gocdk/launch.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (vangent) [reply internal/cmd/gocdk/launch.go]: I guess so, but it makes it less readable since you're basically aliasing a bunch of things -- I have to keep track of the fact that `dockerEnv` is really `pctx.env`. There are so many different envs and dirs floating around that extra names for existing ones just adds to the confusion.

File: internal/cmd/gocdk/launch.go (hunk 1 above)
Comment (vangent) [reply internal/cmd/gocdk/launch.go]: But isn't `newLauncher` used for all launchers as well? I think your comment would make sense to me if there were a separate function being called to construct each launcher, but there's only one (with a switch inside it). So the args to `newLauncher` aren't really specific to a launcher type are they?

File: internal/cmd/gocdk/launch.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (zombiezen) [reply internal/cmd/gocdk/launch.go]: Now I understand. I'm worried about the breaking of convention: when specifying Docker image name + tag, the `name[:tag]` form is prevailing. I've never seen a system that accepts `[name:]tag` or `[name]:tag` as input. Even `docker build -t` (the "tag" flag) takes in the `name[:tag]` form. We could separate this out as two different flags, but that can have surprising effects for the user.

It may be nice for `gocdk deploy` to have such a tag flag, since that would affect both `build` and `launch`, but I don't know if that's in scope for this PR. Can we open an issue for this and proceed from there?

File: internal/cmd/gocdk/launch.go (hunk 1 above)
Comment (zombiezen) [reply internal/cmd/gocdk/launch.go]: It is broadly speaking helpful for the interface split, especially if we move forward with #2046 as we discussed offline.

File: internal/cmd/gocdk/launch.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (zombiezen) [reply internal/cmd/gocdk/launch.go]: Ack. I'd like to address this as part of #2046.

File: internal/cmd/gocdk/launch.go (hunk 2 above)
Comment (zombiezen) [reply internal/cmd/gocdk/launch.go]: Added log statement. Unlaunch/stop is likely going to be too provider-specific since you need to give a handle to a previous deployment. Honestly, I think the more common situation will be that folks just run `gocdk launch -image=MYIMAGE:OLDTAG` to do a rollback.

File: internal/cmd/gocdk/launch.go (hunk 10 above)
Comment (zombiezen) [reply internal/cmd/gocdk/launch.go]: Done.

File: internal/cmd/gocdk/launch.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (zombiezen) [reply internal/cmd/gocdk/launch.go]: Did a little more digging on this, since I want to consider the effect on `gocdk build`. The definitive grammar for a Docker image reference is here: https://github.com/docker/distribution/blob/master/reference/reference.go

So in examining that, `:tag` would never be an acceptable image reference, so we could permit that short form to take the name from the project Dockerfile. Implemented that, WDYT?

File: internal/cmd/gocdk/launch.go (hunk 1 above)
Comment (vangent) [reply internal/cmd/gocdk/launch.go]: I'm fine with delaying this until then. Can you add an explicit TODO about figuring out which arguments should be passed to create the launcher, vs which arguments should be passed to `launcher.Launch`? It's not obvious, especially since `Launch` is likely only going to be called once.

File: internal/cmd/gocdk/launch.go (hunk 5 above)
Comment (vangent) [reply internal/cmd/gocdk/launch.go]: Do you mind removing it for now? If you feel like it needs to be added to make the Cloud Run launcher work well, you can reintroduce it in that PR, but at this point I think it just adds a level of indirection that's hard to follow and doesn't provide much value.

Comment (zombiezen) [reply internal/cmd/gocdk/launch.go]: Removed.
